Add a credentials document to the docs (#98)
ansible/awx-plugins: docs/credentials.rst

    :alt: Credentials - create machine credential form

Machine credentials have several attributes that may be configured:

- **Username**: The username to be used for SSH authentication.
- **Password**: The actual password to be used for SSH authentication. This password will be stored encrypted in the database, if entered. Alternatively, you can configure AWX to ask the user for the password at launch time by selecting :guilabel:`Prompt on launch`. In these cases, a dialog opens when the job is launched, promoting the user to enter the password and password confirmation.
- **SSH Private Key**: Copy or drag-and-drop the SSH private key for the machine credential.
- **Private Key Passphrase**: If the SSH Private Key used is protected by a password, you can configure a Key Password for the private key. This password will be stored encrypted in the database, if entered. Alternatively, you can configure AWX to ask the user for the password at launch time by selecting :guilabel:`Prompt on launch`. In these cases, a dialog opens when the job is launched, prompting the user to enter the password and password confirmation.

> - **Privilege Escalation Method**: Specifies the type of escalation privilege to assign to specific users. This is equivalent to specifying the ``--become-method=BECOME_METHOD`` parameter, where ``BECOME_METHOD`` could be any of the typical methods described below, or a custom method you've written. Begin entering the name of the method, and the appropriate name auto-populates.

Reviewer: We can mark the variable part of the example CLI flag as follows:
```suggestion
- **Privilege Escalation Method**: Specifies the type of escalation privilege to assign to specific users. This is equivalent to specifying the :samp:`--become-method={BECOME_METHOD}` parameter, where ``BECOME_METHOD`` could be any of the typical methods described below, or a custom method you've written. Begin entering the name of the method, and the appropriate name auto-populates.
```

- **Privilege Escalation Username** field is only seen if an option for privilege escalation is selected. Enter the username to use with escalation privileges on the remote system.
- **Privilege Escalation Password**: field is only seen if an option for privilege escalation is selected. Enter the actual password to be used to authenticate the user via the selected privilege escalation type on the remote system. This password will be stored encrypted in the database, if entered. Alternatively, you may configure AWX to ask the user for the password at launch time by selecting :guilabel:`Prompt on launch`. In these cases, a dialog opens when the job is launched, promoting the user to enter the password and password confirmation.

.. note::
   Sudo Password must be used in combination with SSH passwords or SSH Private Keys, since AWX must first establish an authenticated SSH connection with the host prior to invoking sudo to change to the sudo user.

.. warning::

>    Credentials which are used in *Scheduled Jobs* must not be configured as ":guilabel:`Prompt on launch`".

Reviewer: ```suggestion
   Credentials which are used in *Scheduled Jobs* must not be configured as :guilabel:`Prompt on launch`.
```

.. _ug_credentials_network:

Network
~~~~~~~

Select the Network credential type **only** if you are using a ``local`` connection with ``provider`` to use Ansible networking modules to connect to and manage networking devices. When connecting to network devices, the credential type must match the connection type:

- For ``local`` connections using ``provider``, credential type should be :guilabel:`Network`
- For all other network connections (``httpapi``, ``netconf``, and ``network_cli``), credential type should be :guilabel:`Machine`

For an overview of connection types available for network devices, refer to `Multiple Communication Protocols <https://docs.ansible.com/ansible/devel/network/getting_started/network_differences.html#multiple-communication-protocols>`_.

AWX uses the following environment variables for Network credentials and are fields prompted in the user interface:

> .. code-block:: console

Reviewer: What is this? It doesn't look like console output or a shell prompt.

> .. code-block:: console
> 
>   ANSIBLE_NET_USERNAME
>   ANSIBLE_NET_PASSWORD

Reviewer: Try this
```suggestion
.. envvar:: ANSIBLE_NET_USERNAME

   Username to authenticate as when managing networking hardware.

   :Example: ``admin``

.. envvar:: ANSIBLE_NET_PASSWORD

   Password to authenticate with when managing networking hardware.

   :Example: ``Sup3R$s3cuRE``
```

- A token associated with that service account (OpenShift or Kubernetes Bearer Token)
- A CA certificate associated with the cluster

Create a Service Account in an OpenShift cluster (or K8s) in order to be used to run jobs in a container group via AWX. After the Service Account is created, its credentials are provided to AWX in the form of an OpenShift or Kubernetes API bearer token credential. Below describes how to create a service account and collect the needed information for configuring AWX.

To configure AWX:

1. To create a service account, you may use the following service account and modify it as needed to obtain the above credentials:

> .. code-block:: console

Reviewer: ```suggestion
.. code-block:: yaml
```

    ---
    apiVersion: v1
    kind: ServiceAccount
    metadata:
      name: containergroup-service-account
      namespace: containergroup-namespace
    ---
    kind: Role
    apiVersion: rbac.authorization.k8s.io/v1
    metadata:
      name: role-containergroup-service-account
      namespace: containergroup-namespace
    rules:
      - apiGroups: [""]
        resources: ["pods"]
        verbs: ["get", "list", "watch", "create", "update", "patch", "delete"]
      - apiGroups: [""]
        resources: ["pods/log"]
        verbs: ["get", "list", "watch", "create", "update", "patch", "delete"]
      - apiGroups: [""]
        resources: ["pods/attach"]
        verbs: ["get", "list", "watch", "create", "update", "patch", "delete"]
    ---
    kind: RoleBinding
    apiVersion: rbac.authorization.k8s.io/v1
    metadata:
      name: role-containergroup-service-account-binding
      namespace: containergroup-namespace
    subjects:
      - kind: ServiceAccount
        name: containergroup-service-account
        namespace: containergroup-namespace
    roleRef:
      kind: Role
      name: role-containergroup-service-account

>       apiGroup: rbac.authorization.k8s.io

Reviewer: Might be a good idea to keep this in a file that is linted.

Author: It's not causing CI to break right now. Let's do that in another PR if you think that's a better approach.

Reviewer: Only because the problems are now hidden. They should be fixed instead of being swept under the rug.

Reviewer: I can help with this if need be.

Author: Thanks, I appreciate the recommendation. I will open a separate PR for this. I need to get moving along to inventories, but as soon as I get those wrapped up, I'll revisit this all by itself.

Reviewer: I added linting to embedded code blocks just a few minutes ago. Should've done that long ago.

> 2. Apply the configuration from ``containergroup-sa.yml``:

Reviewer: ```suggestion
2. Apply the configuration from :file:`containergroup-sa.yml`:
```

-  **Username**: The username to use to connect to it.
-  **Password**: The password to use to connect to it.
-  **Oauth Token**: If username and password is not used, provide an OAuth token to use to authenticate.

.. _ug_credentials_scm:

Source Control
~~~~~~~~~~~~~~

SCM (source control) credentials are used with Projects to clone and update local source code repositories from a remote revision control system such as Git or Subversion.

|Credentials - create SCM credential|

.. |Credentials - create SCM credential| image:: _static/images/credentials-create-scm-credential.png
    :alt: Credentials - create SCM credential form

Source Control credentials have several attributes that may be configured:

-  **Username**: The username to use in conjunction with the source control system.
-  **Password**:  The password to use in conjunction with the source control system.
-  **SCM Private Key**: Copy or drag-and-drop the actual SSH Private Key to be used to authenticate the user to the source control system via SSH.
-  **Private Key Passphrase**: If the SSH Private Key used is protected by a passphrase, you may configure a Key Passphrase for the private key.

.. note::

>     Source Control credentials cannot be configured as ":guilabel:`Prompt on launch`".

Reviewer: ```suggestion
    Source Control credentials cannot be configured as :guilabel:`Prompt on launch`.
```

    If you are using a GitHub account for a Source Control credential and you have 2FA (Two Factor Authentication) enabled on your account, you will need to use your Personal Access Token in the password field rather than your account password.

.. _ug_credentials_vault:

Vault
~~~~~

Selecting this credential type enables synchronization of inventory with Ansible Vault.

|Credentials - create Vault credential|

.. |Credentials - create Vault credential| image:: _static/images/credentials-create-vault-credential.png
    :alt: Credentials - create Vault credential form

Vault credentials require the **Vault Password** and an optional **Vault Identifier** if applying multi-Vault credentialing.

You may configure AWX to ask the user for the password at launch time by selecting :guilabel:`Prompt on launch`. In these cases, a dialog opens when the job is launched, promoting the user to enter the password and password confirmation.

.. warning::

>     Credentials which are used in *Scheduled Jobs* must not be configured as ":guilabel:`Prompt on launch`".

Reviewer: ```suggestion
    Credentials which are used in *Scheduled Jobs* must not be configured as :guilabel:`Prompt on launch`.
```

- For ``local`` connections using ``provider``, credential type should be :guilabel:`Network`
- For all other network connections (``httpapi``, ``netconf``, and ``network_cli``), credential type should be :guilabel:`Machine`

For an overview of connection types available for network devices, refer to `Multiple Communication Protocols`_.
For an overview of connection types available for network devices, refer to `Multiple Communication Protocols <https://docs.ansible.com/ansible/devel/network/getting_started/network_differences.html#multiple-communication-protocols>`_.

>   .. _`Multiple Communication Protocols`: https://docs.ansible.com/ansible/devel/network/getting_started/network_differences.html#multiple-communication-protocols.

Reviewer: @tvo318 I think the problem here was that the trailing period was being parsed as a part of the URL. Just removing that dot would've fixed the problem.

Author: Ok - I just changed the format because I didn't like the number of lines it took up anyway.

Reviewer: Detached refs are usually better because they don't interrupt plain text in paragraphs as much.

Author: OK, I'll do detached refs next time.

Reviewer: Marking this thread as resolved.
